### ISSUE TITLE
Tracking Issue for `const_heap`

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#79597]
This is a tracking issue for https://github.com/rust-lang/const-eval/issues/20
The feature gate for the issue is `#![feature(const_heap)]`.
The general design can be found [here](https://hackmd.io/h2O2vkj3RimrBTfm9hvZWA#AllocRef-allocGlobal-and-allocConstGlobal)

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [ ] Implement https://github.com/rust-lang/const-eval/issues/20 experimentally
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

* Do we want to allow the regular `Global` allocator? This would give us post-monomorpization errors. More details can be found in https://hackmd.io/h2O2vkj3RimrBTfm9hvZWA#Transient-heap-allocations-via-Global

### Implementation history

* #79594 adds the base intrinsic for telling the miri-engine to create a virtual heap allocation

**Comment by RalfJung** [comment]
As [discussed here](https://github.com/rust-lang/rust/pull/79594#discussion_r533279506), this feature makes `const fn` observably impure: `let x = f(); (x, x)` is no longer always equivalent to `(f(), f())`. We are at a crossroads for the language here: if we want to support heap allocations in `const fn`, we have to either introduce some kind of effect system or say goodbye to the idea that `const fn` denotes a form of purity that would be useful to unsafe code.

**Comment by RalfJung** [comment]
Btw, given how far-reaching the consequences of this feature are, I wonder if doing this without an RFC is really appropriate. But it could make sense to experiment some more with what can be done for transient vs. non-transient allocations, and how `Global` comes in, so that we have a better idea of the design space.

@oli-obk something is wrong with "implementation history", it links to the same PR twice.

**Comment by oli-obk** [comment]
The experimentation that @vn-ki has been doing has exposed various problems that aren't even related to heap stuff itself, but are prerequisites. I believe we should keep on doing this experimentation, just to find out what all the problems are and how a possible implementation could look. Doing this in the abstract has not allowed us to realize what the dependency graph actually is and I will be updating the skill tree to reflect that as we gain more information. Some examples off the top of my head:

* generic types are considered having interior mutability within constants and thus forbid short lived immutable references to them
* trait bounds on methods only work if the bounds are on the method and not if they are on the containing impl block
* the existing alloc code needs support for closures or at least `impl const Fn for const fn()`
* just needs an impl: use fallible allocations for `Allocation` to make sure that we report an error instead of crashing rustc with OOM (this can be done on stable via the use of large arrays)

**Comment by RalfJung** [comment]
The first three points don't seem really related to heap allocations though?

**Comment by oli-obk** [comment]
They are not, they are prerequisites for `Box::new_in` and whatever things are called by that.

**Comment by RalfJung** [comment]
> generic types are considered having interior mutability within constants and thus forbid short lived immutable references to them

Is there a separate issue for this? I think this is closely related to some of the other issues that came up over the last year -- where we have a restriction that applies to *the final value of all consts*, but be unnecessarily apply it even during the course of the computation -- since we have no good way to otherwise ensure that the final value is okay. `&mut` references have basically the same problem.

**Comment by oli-obk** [comment]
There is not. I have a local branch that generalizes to all such checks, so we'll be able to just check them on the final value

**Comment by RalfJung** [comment]
For associated consts, I assume that's a post-monomorphization error then?

(I know this is off-topic here, that's why I asked if there is a better place.^^)

**Comment by oli-obk** [comment]
No... this is a static check that is overzealous in the presence of generics and such.

**Comment by RalfJung** [comment]
Okay, looking forward to it. :) Doing a write-up first before finishing the implementation could also help (but we can also do that later if you prefer... I just think we should do it before landing; "writeup later" hasn't worked out well for us in the past^^).

**Comment by arnabanimesh** [comment]
I personally think that heap allocation at compile time should be done independently. It is a bad idea to include it in `const fn`. `const fn` should not have heap allocations since it increases the code analysis complexity considerably.

I would suggest independently create `const` using heap allocation before starting compilations. Using a separate file known as `consts.rs` in the `root` would be one way of doing it and instantiating the same using `Cargo.toml` (as in case of `build.rs`). However, unlike `build.rs`, it should be able to accept multiple files as it will help in organizing constants for multiple binaries as multiple files. It can have all the rust language features but it should have compiler checks that ensure that at the end only `const` variables can be created and each of them has to be assigned to at least one binary with the same name. Generally, it is `main`, but sometimes multiple `bins` with their own names are also possible.

~It can be done similarly using `build.rs` itself but additional language structs need to be implemented and I don't find it appealing.~

It is just like running `configure` before `make`. It should be doable.

**Comment by oli-obk** [comment]
What you describe (and as you noted), can already be done via `build.rs`. This issue is specifically tracking heap at compile-time, we are very far from stabilization, and I'll add your concerns to the main post. Please be aware that this is a frequently asked for feature, so there is different desires for language design that need to be balanced against each other.

That said, at the current design phase, you can statically know wheter a `const fn` does heap allocations internally, because it must have a `A: AllocRef` generic parameter. If it does not have this, it can only use the `ConstHeap` heap allocator, which at present cannot be used at runtime at all (there is no runtime implementation for the relevant intrinsic, so you'd get a (right now very bad) compile-time error).

**Comment by petrochenkov** [comment]
How far is the implementation state from being able to construct a `FxHashMap` at compile time?

It would be pretty useful for string interner in the compiler itself given that available static alternatives like `phf` appear to be slower  in this case (https://github.com/rust-lang/rust/pull/80420#issuecomment-752724455).

**Comment by oli-obk** [comment]
Very far. This requires not just const heap, but also all kinds of trait methods to be `const`, which has its own blockers (I should really update the const skill tree with all the new information I have now. I guess since we are on nightly we can go quite crazy with feature gates, but we'd first have to implement appropriate const stability attributes for trait impls so we can make trait impls in libcore/libstd `const`

**Comment by nouzu** [comment]
I've implemented a [const-allocatable Vec](https://github.com/nouzu/const_collections/blob/main/src/vec.rs) using this feature. However, according to [my benchmark](https://github.com/nouzu/const_collections/blob/main/benches/vec.rs), its push operation is 15x slower than libstd Vec's. 

My implementation:
```rust
pub struct Vec<T, const N: usize> {
    len: usize,
    mem: NonNull<T>
}

#[inline(always)]
pub const fn push(&mut self, val: T) -> Result<(), T> {
    if self.len == N {
        Err(val)
    } else {
        Ok(unsafe { self.push_unchecked(val) })
    }
}

#[inline(always)]
pub const unsafe fn push_unchecked(&mut self, val: T) {
    self.mem.as_ptr()
        .offset(self.len as _)
        .write(val);

    self.len += 1;
}
```

libstd's implementation:
```rust
pub struct Vec<T, #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global> {
    buf: RawVec<T, A>,
    len: usize,
}

#[inline]
pub fn push_within_capacity(&mut self, value: T) -> Result<(), T> {
    if self.len == self.buf.capacity() {
        return Err(value);
    }
    unsafe {
        let end = self.as_mut_ptr().add(self.len);
        ptr::write(end, value);
        self.len += 1;
    }
    Ok(())
}
```

Benchmark:
```rust
type T = u8;

const L: usize = 4096;

#[bench]
fn std_vec(b: &mut Bencher) {
    let mut vec = Vec::<T>::with_capacity(L);

    b.iter(|| unsafe {
        for i in 0..L {
            vec.push_within_capacity(i as _).unwrap();
        }

        vec.set_len(0);
    });
}

#[bench]
fn const_vec(b: &mut Bencher) {
    static mut VEC: const_collections::vec::Vec<T, L> = const_collections::vec::Vec::new();

    b.iter(|| unsafe {
        for i in 0..L {
            VEC.push(i as _).unwrap();
        }

        VEC.set_len(0);
    });
}
```

In my Cargo.toml I had LTO enabled and codegen-units set to 1. I used `cargo +nightly bench -Zbuild-std --target=x86_64-unknown-linux-gnu` to run the benchmark. This yielded the following result:
```
test const_vec ... bench:         780 ns/iter (+/- 73)
test std_vec   ... bench:          51 ns/iter (+/- 1)
```
